textile/chat-features.textile

```diff
@@ -9,5 +9,495 @@ jump_to:
 
 h2(#overview). Overview
 
-Section dedicated to new chat feature development. TBC
+This document outlines the feature specification for the Ably Chat product.
 
+The key words "must", "must not", "required", "shall", "shall not", "should", "should not", "recommended",  "may", and "optional" (whether lowercased or uppercased) in this document are to be interpreted as described in "RFC 2119":https://tools.ietf.org/html/rfc2119.
+
+The over-the-wire protocol for Chat, as well as implementation details, may be found on the chat protocol page.
+
+h2(#overview). Chat Specification Version
+
+* @(CHA-V1)@ **Specification Version**: This document defines the Ably chat library features specification ('features spec').
+** @(CHA-V1a)@ The current version of the Chat library feature specification is version @0.1.0@.
+
+h2(#general). General Principles
+
+* @(CHA-GP1)@ As far as is practicable, the implementation details such as underlying Realtime Channels should be hidden from the public API. This allows developers to interact with Chat without having to understand many low-level primitives.
+* @(CHA-GP2)@ The public API should avoid implicit operations as a side-effect to some other operation. For example, adding a subscriber to messages in a Chat Room should not automatically trigger that Room to attach. This is in contrast to the current core SDKs. Avoiding side-effects provides a clean, easy to understand API.
+* @(CHA-GP3)@ Wherever possible, Chat features should be exposed in the public API as properties of their parent. For example, @messages@ would be considered a property of a @room@ object. This allows for greater composability and extensibility in the future.
```

Comment:
Too confusing statement -> 
Chat features should be exposed in the public API as properties of their parent. 
I think terms like
1. Chat feature
2. Chat contributor 
3. parent of chat feature
should be explicitly defined

Reply:
When I look at public API, all of the operations are focused on room which is Atomic element of `chat`.
But not sure why we use `chat feature`, `chat contributor`, this makes spec more confusing : (

Reply:
Instead we could call it `room feature`, `room object` etc

Reply:
> This is a language idiomatic statement and can't be generalized
Why do you think it can’t be generalized? I believe that all OOP languages have some concept of properties (members, attributes).

> Chat feature
> Chat contributor

What is not clear for you in Room Lifecycle where Chat "feature" an "contributor" are explained?

> Parent of chat feature 

Maybe it’s not the best choice of words, but it’s pretty clear that features are properties of an object, right? This object is essentially the parent.

Reply:
I can agree with member or attribute, property has a different meaning in dotnet. They have getters and setters only.

Reply:
@ttypic went through the ably-js public API doc and now I can visualize a bit more by the statement `Parent of chat feature`. But, for someone who doesn't know anything about public API, the statement in itself is very confusing.
It should either be purely technical or written from a business point of view. 
Missing IDL can also be one of the reasons people will face issues understanding the spec.

Reply:
> property has a different meaning in dotnet.

dotnet's property is exactly what we want here 

> Missing IDL can also be one of the reasons people will face issues understanding the spec.

Why do you need IDL? We already have all public interfaces and data structures defined

Reply:
We have it defined, but inconsistent naming makes it difficult to understand. Like `DefaultRoom`, `DefaultMessages` class etc. Only when we look at public API, we understand the use-case

Reply:
`property` is a technical term that is only applicable for dotnet, kotlin etc.  For like java, go languages, member as a generalised term would make more sense.

Reply:
`property` often broadly describes a member that is accessible via getters and setters, whilst it has special meaning in some languages, it's quite a general term. In any case, the C# idiom is exactly what we want here (albeit, private setters).

Reply:
Yeah but it also mentions parent child relationship at the same time,
ParentObject#property, which is way easier to understand. 
I would loved if we list out exact types of contributors that exists and their relationship with parent. 
i.e. in our case, there's only one -> room